### ISSUE TITLE
Incorrect interpretation of 'schedule.every().hour.at("17:58").do(job)'

### ISSUE THREAD

**Issue by helarsen** [dbader/schedule#405]
There are some inconsistency/bug in the interface of this otherwise excellent and useful library .

I will illustrate with an example.

schedule.every().hour.at("17:58").do(job)  

will schedule at HH:MM:SS = XX:58:00 and not as one reasonably could expect HH:MM:SS=XX:17:58
So the interface ignores (HH) part of the entered value and imposes a new part (SS).
I claim, that ignoring as little as possible from the entered value (here nothing) is better than ignore something and imposing something additional - in this case the entered value has no ambiguity and should be taken as it is.

Unfortunately I don't think it is possible to correct this without breaking existing code.

Moreover there are the following inconsistent behaviors:
schedule.every().hour.at("16:17:58").do(job)  
returns 
Invalid schedule.ScheduleValueError: Invalid time format for an hourly job. 
What it should do is interpret the time as HH:MM:SS = XX:17:58, i.e. ignore the (HH) part. This would be logical behavior for an hourly task

Along the same lines: 
schedule.every().minute.at("16:17:58").do(job)  
returns 
Invalid schedule.ScheduleValueError: Invalid time format for an minutely job.
What it should do is interpret the time as HH:MM:SS = XX:YY:58, i.e. ignore the (HH:MM) part. This would be logical behavior for an minutely task

Thanks for considering this.

**Comment by SijmenHuizenga** [comment]
Hi @helarsen, thanks for sharing your concerns. These are my two cents on the topics

> `schedule.every().hour.at("17:58").do(job)` will schedule at HH:MM:SS = XX:58:00 and not as one reasonably could expect HH:MM:SS=XX:17:58 So the interface ignores (HH) part of the entered value and imposes a new part (SS).

Good point! I agree it is weird that schedule accepts `hour.at("17:58")` and interprets it as `hour.at(":58")`. You are correct that this kind of ambiguity can lead to problems. Weather or not we should publish a breaking change to fix this... I'm not sure. Let's track this bug in #286 (and pr #290).

> Moreover there are the following inconsistent behaviors: `schedule.every().hour.at("16:17:58").do(job)` returns `Invalid schedule.ScheduleValueError: Invalid time format for an hourly job.`

This error seems correct to me because the time format and the time unit imply a different thing:
* `schedule.every().hour` implies the job should run every hour.
* `.at("16:17:58")` implies the job should run only once a day.

To prevent ambiguity schedule throws an error. Schedule wants either `schedule.every().day.at("16:17:58")` or `schedule.every().hour.at("17:58")`

> What it should do is interpret the time as HH:MM:SS = XX:YY:58, i.e. ignore the (HH:MM) part. This would be logical behavior for an minutely task

I think schedule should follow all the instructions it is given. If schedule ignores certain parameters the code becomes ambiguous. In this case, what instructions will schedule prioritize? Should it ignore `.seconds` and run daily or ignore the `HH:MM` part? Users will have different assumptions and opinions so the best way is to require people to write unambiguous code.

**Comment by helarsen** [comment]
Further to the HH:MM:SS format being invalid time format for 
`schedule.every().hour.at("hh:mm:ss").do(job)`
`schedule.every().minute.at("hh:mm:ss").do(job)`

@SijmenHuizenga argues:
> This error seems correct to me because the time format and the time unit imply a different thing:
>   `schedule.every().hour` implies the job should run every hour.
 >   `.at("16:17:58")` implies the job should run only once a day.
>To prevent ambiguity schedule throws an error. Schedule wants either schedule.every().day.at("16:17:58") or schedule.every().hour.at("17:58")

To this I would argue that the command must be seen and interpreted in is full and not as independent parts so it is not ambiguous.

But let me continue - It is a reasonable assumption that any user should know the meaning of the format: "HH:MM:SS". Right?

If so,
`schedule.every().hour.at("hh:mm:ss").do(job,) `
can only have 2 reasonable interpretations:
A) `HH:MM:SS = XX:mm:ss` (<-  as it is periodic every hour it makes no sense to specify HH i.e it is ignored)
B) `HH:MM:SS = XX:mm:00` 
C) all other are non sense like `HH:MM:SS = XX:hh:mm` where the number in HH position is moved to MM field as one eample.

Interpretation (B) is not logical - "Why throw away well specified information and impose some new default value (SS=00)?" So this option should not be used.

But - worse still, **there is no way to program the `schedule.every().hour.at("-").do(job)` to run at a specific SS second different from 00** - you can only control the MM part of the run-time! I find this implementation is lame (does not function to its full potential) because of these reasons. By the way, that is the reason I tumbled upon this issue in the first place!

_I think a solution should be provided to schedule an hourly task at any SS second instance_. Maybe this is actually a new issue.

A **solution could be to allow the `schedule.every().hour.at("hh:mm:ss").do(job`)`.**
This would solve this issue in an efficient manner which does not break existing code.

To be logical and consistent, this change should also apply to.
`schedule.every().minute.at("hh:mm:ss").do(job)`
This is perfectly non redundant and can only have 1 reasonable interpretation:
A) `HH:MM:SS = xx:yy:ss` (<- as it is periodic every minute it makes no sense to specify MM or HH)
 all other interpretations are non sense like HH:MM:SS = xx:yy:mm (<-removing HH and SS part)

It matches nicely with that `schedule.every().day.at("hh:mm:ss").do(job)` is a valid command.

I hope my arguments are clear and thus convincing.

**Comment by SijmenHuizenga** [comment]
Thank you for your detailed write up @helarsen, your arguments are super clear.

Not being able to schedule a job to run every hour at a specific second is a bug and should be fixed. You are totally right. It has been reported before here #286 and a pr is open here #290.

> I think a solution should be provided to schedule an hourly task at any SS second instance.

Yes, I agree.

> A solution could be to allow the schedule.every().hour.at("hh:mm:ss").do(job)`.

I'm worried that  `every().hour.at("hh:mm:ss").do` can be interpreted in multiple ways:

1. run daily at hh:mm:ss
2. run hourly at mm:ss

A user has to _know_ that a the `.hour` takes priority over the hour-value specified in the `.at`.  You assume 2 (the HH is ignored) but I don't think everyone will. 

```
schedule.every().minute.at("12:16:42").do(job)
```

I can imagine people reading this line of code and get confused... people will be asking what the `12` and `16` value do. The current syntax:

```
schedule.every().minute.at(":42").do(job)
```

does not have this problem. Everyone will understand the job has to run every minute at the 42th second.

If I understand you correctly (and please correct me if I'm wrong), you feel the values entered in `.at()` should be treated the same, independently of the interval of the job. I think it is even more important that the code one writes using this library is understandable. Therefore I don't think we should add `hh:mm:ss` for `.hour` and `.second` jobs.

**Comment by helarsen** [comment]
hi @SijmenHuizenga,
Thanks for commenting on my proposals and pointing out that the problem `schedule.every().hour.at("-").do(job)` at any SS is a know problem and supporting that a resolution is required.
You ask
> If I understand you correctly (and please correct me if I'm wrong), you feel the values entered in .at() should be treated the same, independently of the interval of the job. I think it is even more important that the code one writes using this library is understandable. Therefore I don't think we should add hh:mm:ss for .hour and .second jobs.

Yes you are correct that is what I find.
"Note 1 - you mean .minute and not .second of course"
> I can imagine people reading this line of code and get confused... people will be asking what the 12 and 16 value do. The current syntax: `schedule.every().minute.at(":42").do(job)`

Ok, I will not claim that it cannot confuse but as I wrote, if you consider the entire command and what you are asking for `.every().hour` then it _should_ be clear

The manual says
> Parameters: 
time_str – A string in one of the following formats: HH:MM:SS, HH:MM, :MM, :SS. The format must make sense given how often the job is repeating; for example, a job that repeats every minute should not be given a string in the form HH:MM:SS. The difference between :MM and :SS is inferred from the selected time-unit (e.g. every().hour.at(‘:30’) vs. every().minute.at(‘:30’)).

Should in stead read 
> Parameters: 
time_str – A string with the following valid entries depending on context
>- `every().day.at(‘HH:MM:SS’)`: 

>- `every().hour.at(‘HH:MM:SS’)` -> the HH: part is ignored
`every().hour.at(‘:MM:SS’)`  ->   _Note 2_
`every().hour.at(‘MM:SS’)`  ->   _Note 2_
`every().hour.at(‘:MM’)`        -> equivalent to `every().hour.at(‘MM:00’)`  or `every().hour.at(‘HH:MM:00’)`

>- `every().minute.at(‘HH:MM:SS’)` -> the HH:MM part is ignored
`every().minute.at(‘:MM:SS’)` -> the :MM part is ignored
`every().minute.at(‘MM:SS’)` -> the MM part is ignored
`every().minute.at(‘:SS’)`


My claim is that documentation in this way - listing all allowed options and defaults is much more clear than the current documentation.
Note 2: Notice how the current interface is lacking the MM:SS or :MM:SS formats. Probably the backgound of this oversight.

(gosh  - it is hard to get the markup typesetting right)

I don't expect to enter much more into this case as I have found a workaround for my particular application and I think my points are clear. Take from it what you want. Have a nice day. /henning
